Group instructor autocomplete

cxcorp opened the pull request:
Backend dependencies merged already.

Splits GroupManagementPage.js's components into their own files and adds an autocompleted user selection component which is used to select the instructor of a group by name.

cxcorp commented:
Requires `npm install` because of new dependency to `react-select`.

Ha3a commented:
Looking good, some padding around buttons or instructor search field would be a nice addition

cxcorp commented:
Actual diff after splitting files (361e7e3), generated with `git diff 361e HEAD src/ e2e package.json`:

```diff
diff --git a/e2e/cypress/integration/groupManagementPage.spec.js b/e2e/cypress/integration/groupManagementPage.spec.js
index 7459252..6824106 100644
--- a/e2e/cypress/integration/groupManagementPage.spec.js
+++ b/e2e/cypress/integration/groupManagementPage.spec.js
@@ -13,7 +13,21 @@ describe('Group Management Page', () => {
       cy.get('.topic-menu-item-no__1').click()
       cy.get('.create-group-form__name').type('Ryhmä A')
       cy.get('.create-group-form__student-input').type('012345678, 012345688')
-      cy.get('.create-group-form__instructor').type('012345698')
+
+      cy.get('.create-group-form__instructor').click()
+
+      cy.get('.create-group-form__instructor')
+        .find('input')
+        .type('Angela', { force: true })
+
+      cy.get('.create-group-form__instructor').find(
+        '.create-group-form__instructor__option'
+      )
+
+      cy.get('.create-group-form__instructor')
+        .find('input')
+        .type('{enter}', { force: true })
+
       cy.get('.create-group-submit').click()
       cy.get('.notification').should('have.text', 'Group saved!')
     })
@@ -82,9 +96,20 @@ describe('Group Management Page', () => {
     })
 
     it('changes instructor', () => {
+      cy.get('.edit-group-no__1__instructor').click()
+
       cy.get('.edit-group-no__1__instructor')
-        .clear()
-        .type('012345678')
+        .find('input')
+        .type('{backspace}', { force: true })
+
+      cy.get('.edit-group-no__1__instructor')
+        .find('input')
+        .type('Testaaja', { force: true })
+
+      cy.get('.edit-group-no__1__instructor')
+        .find('input')
+        .type('{enter}', { force: true })
+
       cy.get('.edit-group-no__1__save-button').click()
       cy.get('.notification').should('have.text', 'Group updated!')
       cy.get('.view-group-no__1__container .group-instructor').should(
diff --git a/package.json b/package.json
index 491c2d2..f5f1f18 100644
--- a/package.json
+++ b/package.json
@@ -21,6 +21,7 @@
     "react-redux": "^5.1.1",
     "react-router-dom": "^4.3.1",
     "react-scripts": "^2.1.3",
+    "react-select": "^2.4.1",
     "redux": "^4.0.1",
     "redux-thunk": "^2.3.0"
   },
diff --git a/src/components/GroupManagementPage/AutocompletedUserSelect.js b/src/components/GroupManagementPage/AutocompletedUserSelect.js
new file mode 100644
index 0000000..d30e582
--- /dev/null
+++ b/src/components/GroupManagementPage/AutocompletedUserSelect.js
@@ -0,0 +1,76 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import AsyncSelect from 'react-select/lib/Async'
+
+import autocomplete from '../../services/autocomplete'
+
+/** @typedef {{ student_number: string, first_names: string, last_name: string }} AutocompleteResult */
+
+const MIN_CHARS = 2
+
+/** @param {AutocompleteResult} option */
+const getOptionLabel = (option) => `${option.first_names} ${option.last_name}`
+
+/** @param {AutocompleteResult} option */
+const getOptionValue = (option) => option.student_number
+
+const AutocompletedUserSelect = ({
+  className,
+  classNamePrefix,
+  selectedUser,
+  onSelectedUserChange,
+  defaultUser
+}) => {
+  /** @param {AutocompleteResult} selectedOption */
+  const handleChange = (selectedOption, { action }) => {
+    if (action === 'clear') {
+      // selectedOption is null
+      onSelectedUserChange(null)
+    } else if (action === 'select-option') {
+      // selectedOption is one of the result objects returned by the
+      // autocomplete service
+      onSelectedUserChange(selectedOption)
+    }
+  }
+
+  const handleLoadOptions = async (inputValue) => {
+    if (inputValue.length < MIN_CHARS) {
+      return
+    }
+
+    return await autocomplete.findUsersByPartialName(inputValue)
+  }
+
+  return (
+    <AsyncSelect
+      cacheOptions
+      defaultOptions
+      isClearable
+      value={selectedUser}
+      loadOptions={handleLoadOptions}
+      onChange={handleChange}
+      getOptionValue={getOptionValue}
+      getOptionLabel={getOptionLabel}
+      defaultValue={defaultUser}
+      placeholder="Search by name"
+      className={className}
+      classNamePrefix={classNamePrefix}
+    />
+  )
+}
+
+const autosuggestResultShape = PropTypes.shape({
+  student_number: PropTypes.string.isRequired,
+  first_names: PropTypes.string.isRequired,
+  last_name: PropTypes.string.isRequired
+})
+
+AutocompletedUserSelect.propTypes = {
+  className: PropTypes.string,
+  classNamePrefix: PropTypes.string,
+  selectedUser: autosuggestResultShape,
+  defaultUser: autosuggestResultShape,
+  onSelectedUserChange: PropTypes.func.isRequired
+}
+
+export default AutocompletedUserSelect
diff --git a/src/components/GroupManagementPage/ConfigurationSelect.js b/src/components/GroupManagementPage/ConfigurationSelect.js
index ab2aa26..8d552e8 100644
--- a/src/components/GroupManagementPage/ConfigurationSelect.js
+++ b/src/components/GroupManagementPage/ConfigurationSelect.js
@@ -4,9 +4,6 @@ import { connect } from 'react-redux'
 import Select from '@material-ui/core/Select'
 import MenuItem from '@material-ui/core/MenuItem'
 
-import topicListPageActions from '../../reducers/actions/topicListPageActions'
-import adminPageActions from '../../reducers/actions/adminPageActions'
-import * as notificationActions from '../../reducers/actions/notificationActions'
 import groupManagementActions from '../../reducers/actions/groupManagementActions'
 
 const ConfigurationSelect = ({
@@ -33,34 +30,13 @@ const ConfigurationSelect = ({
   )
 }
 
-// Todo: these were copypasted from GroupManagementPage - figure out what props
-//       are actually used
 const mapStateToPropsForm = (state) => ({
-  groupName: state.groupPage.groupName,
-  students: state.groupPage.students,
-  groupTopicID: state.groupPage.groupTopicID,
-  groupInstructorID: state.groupPage.groupInstructorID,
   groupConfigurationID: state.groupPage.groupConfigurationID,
-  groups: state.groupPage.groups,
-  topics: state.topicListPage.topics,
-  configurations: state.adminPage.configurations,
-  users: state.groupPage.users
+  configurations: state.adminPage.configurations
 })
 
 const mapDispatchToPropsForm = {
-  onNameChangeForm: groupManagementActions.updateCreateGroupFormName,
-  onStudentFormChange: groupManagementActions.updateStudentsForm,
-  onTopicSelectChange: groupManagementActions.updateGroupTopicID,
-  onInstructorIdChange: groupManagementActions.updateGroupInstructorID,
-  onConfigurationChange: groupManagementActions.updateGroupConfigurationID,
-  deleteFromGroupAction: groupManagementActions.deleteFromGroup,
-  updateExistingGroup: groupManagementActions.updateExistingGroup,
-  createGroupSuccsess: groupManagementActions.createGroupSuccsess,
-  deleteGroup: groupManagementActions.deleteGroup,
-  updateTopicsForm: topicListPageActions.updateTopics,
-  updateConfigurations: adminPageActions.updateConfigurations,
-  setError: notificationActions.setError,
-  setSuccess: notificationActions.setSuccess
+  onConfigurationChange: groupManagementActions.updateGroupConfigurationID
 }
 
 export default connect(
diff --git a/src/components/GroupManagementPage/GroupCreationForm.css b/src/components/GroupManagementPage/GroupCreationForm.css
new file mode 100644
index 0000000..ea6a806
--- /dev/null
+++ b/src/components/GroupManagementPage/GroupCreationForm.css
@@ -0,0 +1,4 @@
+.form-input__label,
+.form-input__control {
+  padding-bottom: 1rem;
+}
diff --git a/src/components/GroupManagementPage/GroupCreationForm.js b/src/components/GroupManagementPage/GroupCreationForm.js
index decd31a..d23c4c6 100644
--- a/src/components/GroupManagementPage/GroupCreationForm.js
+++ b/src/components/GroupManagementPage/GroupCreationForm.js
@@ -7,19 +7,19 @@ import Button from '@material-ui/core/Button'
 
 import groupManagementService from '../../services/groupManagement'
 
-import topicListPageActions from '../../reducers/actions/topicListPageActions'
-import adminPageActions from '../../reducers/actions/adminPageActions'
 import * as notificationActions from '../../reducers/actions/notificationActions'
 import groupManagementActions from '../../reducers/actions/groupManagementActions'
 
 import TopicSelect from './TopicSelect'
+import AutocompletedUserSelect from './AutocompletedUserSelect'
+import './GroupCreationForm.css'
 
 const FormInput = ({ label, children }) => (
   <tr>
-    <td>
+    <td className="form-input__label">
       <Typography variant="caption">{label}</Typography>
     </td>
-    <td>{children}</td>
+    <td className="form-input__control">{children}</td>
   </tr>
 )
 
@@ -54,21 +54,27 @@ const saveGroup = async (event, props) => {
     groupName,
     students,
     groupTopicID,
-    groupInstructorID,
+    groupInstructor,
     groupConfigurationID
   } = props
 
   const splitStudents = students
-    .split(/[;,\n]/)
-    .map((student) => student.trim())
+    .split(/[;, \n]/)
     .filter((str) => !!str)
+    .map((student) => {
+      if (student.length < 9) {
+        return '0' + student.trim()
+      } else {
+        return student.trim()
+      }
+    })
 
   try {
     const createdGroup = await groupManagementService.create({
       name: groupName,
       topicId: groupTopicID,
       configurationId: groupConfigurationID,
-      instructorId: groupInstructorID,
+      instructorId: groupInstructor ? groupInstructor.student_number : '',
       studentIds: splitStudents
     })
     props.createGroupSuccsess(createdGroup)
@@ -84,16 +90,16 @@ const GroupCreationForm = ({
   onNameChangeForm,
   students,
   onStudentFormChange,
-  groupInstructorID,
-  onInstructorIdChange,
+  groupInstructor,
+  onInstructorChange,
   topics,
   onTopicSelectChange,
+
   groupTopicID,
   groupConfigurationID,
   createGroupSuccsess,
   setSuccess,
-  setError,
-  clearNotifications
+  setError
 }) => {
   return (
     <div>
@@ -103,12 +109,11 @@ const GroupCreationForm = ({
             groupName,
             students,
             groupTopicID,
-            groupInstructorID,
+            groupInstructor,
             groupConfigurationID,
             createGroupSuccsess,
             setSuccess,
-            setError,
-            clearNotifications
+            setError
           })
         }
       >
@@ -136,10 +141,11 @@ const GroupCreationForm = ({
             </FormInput>
 
             <FormInput label="Instructor">
-              <NameInput
-                value={groupInstructorID}
-                onChange={onInstructorIdChange}
-                inputProps={{ className: 'create-group-form__instructor' }}
+              <AutocompletedUserSelect
+                className="create-group-form__instructor"
+                classNamePrefix="create-group-form__instructor"
+                selectedUser={groupInstructor}
+                onSelectedUserChange={onInstructorChange}
               />
             </FormInput>
           </tbody>
@@ -158,32 +164,21 @@ const GroupCreationForm = ({
   )
 }
 
-// Todo: these were copypasted from GroupManagementPage - figure out what props
-//       are actually used
 const mapStateToPropsForm = (state) => ({
   groupName: state.groupPage.groupName,
   students: state.groupPage.students,
   groupTopicID: state.groupPage.groupTopicID,
-  groupInstructorID: state.groupPage.groupInstructorID,
+  groupInstructor: state.groupPage.groupInstructor,
   groupConfigurationID: state.groupPage.groupConfigurationID,
-  groups: state.groupPage.groups,
-  topics: state.topicListPage.topics,
-  configurations: state.adminPage.configurations,
-  users: state.groupPage.users
+  topics: state.topicListPage.topics
 })
 
 const mapDispatchToPropsForm = {
   onNameChangeForm: groupManagementActions.updateCreateGroupFormName,
   onStudentFormChange: groupManagementActions.updateStudentsForm,
   onTopicSelectChange: groupManagementActions.updateGroupTopicID,
-  onInstructorIdChange: groupManagementActions.updateGroupInstructorID,
-  onConfigurationChange: groupManagementActions.updateGroupConfigurationID,
-  deleteFromGroupAction: groupManagementActions.deleteFromGroup,
-  updateExistingGroup: groupManagementActions.updateExistingGroup,
+  onInstructorChange: groupManagementActions.updateGroupInstructor,
   createGroupSuccsess: groupManagementActions.createGroupSuccsess,
-  deleteGroup: groupManagementActions.deleteGroup,
-  updateTopicsForm: topicListPageActions.updateTopics,
-  updateConfigurations: adminPageActions.updateConfigurations,
   setError: notificationActions.setError,
   setSuccess: notificationActions.setSuccess
 }
diff --git a/src/components/GroupManagementPage/GroupManagementForm.js b/src/components/GroupManagementPage/GroupManagementForm.js
index 85f6817..fb58921 100644
--- a/src/components/GroupManagementPage/GroupManagementForm.js
+++ b/src/components/GroupManagementPage/GroupManagementForm.js
@@ -3,8 +3,6 @@ import { connect } from 'react-redux'
 
 import Paper from '@material-ui/core/Paper'
 
-import topicListPageActions from '../../reducers/actions/topicListPageActions'
-import adminPageActions from '../../reducers/actions/adminPageActions'
 import * as notificationActions from '../../reducers/actions/notificationActions'
 import groupManagementActions from '../../reducers/actions/groupManagementActions'
 
@@ -18,8 +16,7 @@ const GroupManagementForm = ({
   users,
   deleteGroup,
   setSuccess,
-  setError,
-  clearNotifications
+  setError
 }) => {
   const filteredGroups = groups.filter(
     (group) => group.configurationId === groupConfigurationID
@@ -46,7 +43,6 @@ const GroupManagementForm = ({
                         deleteGroup={deleteGroup}
                         setSuccess={setSuccess}
                         setError={setError}
-                        clearNotifications={clearNotifications}
                       />
                     </td>
                   </tr>
@@ -60,32 +56,16 @@ const GroupManagementForm = ({
   )
 }
 
-// Todo: these were copypasted from GroupManagementPage - figure out what props
-//       are actually used
 const mapStateToPropsForm = (state) => ({
-  groupName: state.groupPage.groupName,
-  students: state.groupPage.students,
-  groupTopicID: state.groupPage.groupTopicID,
-  groupInstructorID: state.groupPage.groupInstructorID,
   groupConfigurationID: state.groupPage.groupConfigurationID,
   groups: state.groupPage.groups,
   topics: state.topicListPage.topics,
-  configurations: state.adminPage.configurations,
   users: state.groupPage.users
 })
 
 const mapDispatchToPropsForm = {
-  onNameChangeForm: groupManagementActions.updateCreateGroupFormName,
-  onStudentFormChange: groupManagementActions.updateStudentsForm,
-  onTopicSelectChange: groupManagementActions.updateGroupTopicID,
-  onInstructorIdChange: groupManagementActions.updateGroupInstructorID,
-  onConfigurationChange: groupManagementActions.updateGroupConfigurationID,
   deleteFromGroupAction: groupManagementActions.deleteFromGroup,
-  updateExistingGroup: groupManagementActions.updateExistingGroup,
-  createGroupSuccsess: groupManagementActions.createGroupSuccsess,
   deleteGroup: groupManagementActions.deleteGroup,
-  updateTopicsForm: topicListPageActions.updateTopics,
-  updateConfigurations: adminPageActions.updateConfigurations,
   setError: notificationActions.setError,
   setSuccess: notificationActions.setSuccess
 }
diff --git a/src/components/GroupManagementPage/GroupViewer.js b/src/components/GroupManagementPage/GroupViewer.js
index 896dae5..62428b3 100644
--- a/src/components/GroupManagementPage/GroupViewer.js
+++ b/src/components/GroupManagementPage/GroupViewer.js
@@ -32,7 +32,6 @@ class GroupViewer extends React.Component {
         <SingleGroupEdit
           group={this.props.group}
           toggleEditMode={this.toggleEditMode}
-          deleteGroup={this.props.deletGroup}
         />
       )
     } else {
@@ -46,7 +45,6 @@ class GroupViewer extends React.Component {
             toggleEditMode={this.toggleEditMode}
             setSuccess={this.props.setSuccess}
             setError={this.props.setError}
-            clearNotifications={this.props.clearNotifications}
           />
         </div>
       )
diff --git a/src/components/GroupManagementPage/SingleGroupEdit.js b/src/components/GroupManagementPage/SingleGroupEdit.js
index 34215dd..fd5a7ab 100644
--- a/src/components/GroupManagementPage/SingleGroupEdit.js
+++ b/src/components/GroupManagementPage/SingleGroupEdit.js
@@ -5,30 +5,13 @@ import TextField from '@material-ui/core/TextField'
 import Button from '@material-ui/core/Button'
 import IconButton from '@material-ui/core/IconButton'
 import DeleteIcon from '@material-ui/icons/Delete'
-import Select from '@material-ui/core/Select'
-import MenuItem from '@material-ui/core/MenuItem'
 
 import groupManagementService from '../../services/groupManagement'
-
-import topicListPageActions from '../../reducers/actions/topicListPageActions'
-import adminPageActions from '../../reducers/actions/adminPageActions'
 import * as notificationActions from '../../reducers/actions/notificationActions'
 import groupManagementActions from '../../reducers/actions/groupManagementActions'
 
-const TopicSelect = ({ topics, onTopicSelectChange, groupTopicID }) => {
-  return (
-    <Select
-      value={groupTopicID}
-      onChange={(e) => onTopicSelectChange(e.target.value)}
-    >
-      {topics.map((topic) => (
-        <MenuItem key={topic.id} value={topic.id}>
-          {topic.content.title}
-        </MenuItem>
-      ))}
-    </Select>
-  )
-}
+import AutocompletedUserSelect from './AutocompletedUserSelect'
+import TopicSelect from './TopicSelect'
 
 const updateCreatedGroup = async (event, props) => {
   event.preventDefault()
@@ -38,14 +21,20 @@ const updateCreatedGroup = async (event, props) => {
     name,
     topicId,
     studentIds,
-    instructorId,
+    instructor,
     configurationId
   } = props.group
 
   const splitStudents = studentIds
-    .split(/[;,\n]/)
-    .map((student) => student.trim())
+    .split(/[;, \n]/)
     .filter((str) => !!str)
+    .map((student) => {
+      if (student.length < 9) {
+        return '0' + student.trim()
+      } else {
+        return student.trim()
+      }
+    })
 
   try {
     const updatedGroup = await groupManagementService.put({
@@ -53,7 +42,7 @@ const updateCreatedGroup = async (event, props) => {
       name: name,
       topicId: topicId,
       configurationId: configurationId,
-      instructorId: instructorId,
+      instructorId: instructor ? instructor.student_number : '',
       studentIds: splitStudents
     })
 
@@ -68,39 +57,6 @@ const updateCreatedGroup = async (event, props) => {
   }
 }
 
-const deleteExistingGroup = async (event, props) => {
-  event.preventDefault()
-
-  const {
-    id,
-    name,
-    topicId,
-    studentIds,
-    configurationId,
-    instructorId
-  } = props.group
-
-  const groupToDelete = {
-    id: id,
-    name: name,
-    topicId: topicId,
-    configurationId: configurationId,
-    instructorId: instructorId,
-    studentIds: studentIds
-  }
-
-  try {
-    await groupManagementService.del({
-      id: id
-    })
-    props.deleteGroup(groupToDelete)
-
-    props.setSuccess('Group deleted!')
-  } catch (e) {
-    props.setError(`Failed to delte! ${e.response.data.error}`)
-  }
-}
-
 class SingleGroupEdit extends React.Component {
   constructor(props) {
     super(props)
@@ -110,7 +66,9 @@ class SingleGroupEdit extends React.Component {
       name: props.group.name,
       topicId: props.group.topicId,
       configurationId: props.group.configurationId,
-      instructorId: props.group.instructorId,
+      instructor: props.users.find(
+        (user) => user.student_number === props.group.instructorId
+      ),
       studentIds: props.group.studentIds.join('\n')
     }
   }
@@ -123,31 +81,54 @@ class SingleGroupEdit extends React.Component {
     this.setState({ studentIds: event.target.value })
   }
 
-  handleInstructorChange = (event) => {
-    this.setState({ instructorId: event.target.value })
+  handleInstructorChange = (newInstructor) => {
+    this.setState({ instructor: newInstructor })
   }
 
   handleTopicChange = (e) => {
     this.setState({ topicId: e })
   }
 
+  handleGroupDelete = async (e) => {
+    e.preventDefault()
+
+    const groupId = this.state.id
+
+    try {
+      await groupManagementService.del({
+        id: groupId
+      })
+      this.props.deleteGroup(groupId)
+
+      this.props.setSuccess('Group deleted!')
+    } catch (e) {
+      this.props.setError(`Failed to delte! ${e.response.data.error}`)
+    }
+  }
+
   render() {
-    const group = this.state
+    const {
+      group,
+      topics,
+      users,
+      updateExistingGroup,
+      toggleEditMode,
+      setSuccess,
+      setError,
+      clearNotifications
+    } = this.props
+
+    const defaultInstructor =
+      group.instructorId &&
+      users.find((user) => user.student_number === group.instructorId)
 
     return (
       <div style={{ allign: 'top' }}>
-        Edit group: {this.props.group.name}
+        Edit group: {group.name}
         <IconButton
           aria-label="Delete"
-          onClick={(event) =>
-            deleteExistingGroup(event, {
-              group,
-              deleteGroup: this.props.deleteGroup,
-              setSuccess: this.props.setSuccess,
-              setError: this.props.setError,
-              clearNotifications: this.props.clearNotifications
-            })
-          }
+          onClick={this.handleGroupDelete}
+          className={`edit-group-no__${group.id}__delete-button`}
         >
           <DeleteIcon fontSize="large" />
         </IconButton>
@@ -159,7 +140,7 @@ class SingleGroupEdit extends React.Component {
         />
         <p>Change topic</p>
         <TopicSelect
-          topics={this.props.topics}
+          topics={topics}
           onTopicSelectChange={this.handleTopicChange}
           groupTopicID={this.state.topicId}
           className="edit-group-form-topic__selector"
@@ -173,69 +154,56 @@ class SingleGroupEdit extends React.Component {
           rows="8"
         />
         <p>Change instructor</p>
-        <TextField
-          inputProps={{ className: `edit-group-no__${group.id}__instructor` }}
-          value={this.state.instructorId || ''}
-          onChange={this.handleInstructorChange}
+        <AutocompletedUserSelect
+          className={`edit-group-no__${group.id}__instructor`}
+          classNamePrefix={`edit-group-no__${group.id}__instructor`}
+          selectedUser={this.state.instructor}
+          defaultUser={defaultInstructor}
+          onSelectedUserChange={this.handleInstructorChange}
         />
-        <Button
-          style={{ marginLeft: '10px', height: '30px', float: 'right' }}
-          color="primary"
-          variant="contained"
-          onClick={(event) =>
-            updateCreatedGroup(event, {
-              group,
-              updateExistingGroup: this.props.updateExistingGroup,
-              toggleEditMode: this.props.toggleEditMode,
-              setSuccess: this.props.setSuccess,
-              setError: this.props.setError,
-              clearNotifications: this.props.clearNotifications
-            })
-          }
-          className={`edit-group-no__${group.id}__save-button`}
-        >
-          Save
-        </Button>
-        <Button
-          style={{ marginLeft: '10px', height: '30px', float: 'right' }}
-          color="primary"
-          variant="contained"
-          onClick={() => this.props.toggleEditMode()}
-          className={`edit-group-no__${group.id}__cancel-button`}
-        >
-          Cancel
-        </Button>
+        <p>
+          <Button
+            style={{ marginLeft: '10px', height: '30px', float: 'right' }}
+            color="primary"
+            variant="contained"
+            onClick={(event) =>
+              updateCreatedGroup(event, {
+                group: this.state,
+                updateExistingGroup,
+                toggleEditMode,
+                setSuccess,
+                setError,
+                clearNotifications
+              })
+            }
+            className={`edit-group-no__${group.id}__save-button`}
+          >
+            Save
+          </Button>
+          <Button
+            style={{ marginLeft: '10px', height: '30px', float: 'right' }}
+            color="primary"
+            variant="contained"
+            onClick={() => toggleEditMode()}
+            className={`edit-group-no__${group.id}__cancel-button`}
+          >
+            Cancel
+          </Button>
+        </p>
       </div>
     )
   }
 }
 
-// Todo: these were copypasted from GroupManagementPage - figure out what props
-//       are actually used
 const mapStateToPropsForm = (state) => ({
-  groupName: state.groupPage.groupName,
-  students: state.groupPage.students,
-  groupTopicID: state.groupPage.groupTopicID,
-  groupInstructorID: state.groupPage.groupInstructorID,
-  groupConfigurationID: state.groupPage.groupConfigurationID,
-  groups: state.groupPage.groups,
   topics: state.topicListPage.topics,
-  configurations: state.adminPage.configurations,
   users: state.groupPage.users
 })
 
 const mapDispatchToPropsForm = {
-  onNameChangeForm: groupManagementActions.updateCreateGroupFormName,
-  onStudentFormChange: groupManagementActions.updateStudentsForm,
   onTopicSelectChange: groupManagementActions.updateGroupTopicID,
-  onInstructorIdChange: groupManagementActions.updateGroupInstructorID,
-  onConfigurationChange: groupManagementActions.updateGroupConfigurationID,
-  deleteFromGroupAction: groupManagementActions.deleteFromGroup,
   updateExistingGroup: groupManagementActions.updateExistingGroup,
-  createGroupSuccsess: groupManagementActions.createGroupSuccsess,
   deleteGroup: groupManagementActions.deleteGroup,
-  updateTopicsForm: topicListPageActions.updateTopics,
-  updateConfigurations: adminPageActions.updateConfigurations,
   setError: notificationActions.setError,
   setSuccess: notificationActions.setSuccess
 }
diff --git a/src/components/GroupManagementPage/SingleGroupView.js b/src/components/GroupManagementPage/SingleGroupView.js
index e759638..409e376 100644
--- a/src/components/GroupManagementPage/SingleGroupView.js
+++ b/src/components/GroupManagementPage/SingleGroupView.js
@@ -13,8 +13,7 @@ const SingleGroupView = ({
   deleteFromGroupAction,
   toggleEditMode,
   setSuccess,
-  setError,
-  clearNotifications
+  setError
 }) => {
   const thisTopic = topics.filter((topic) => topic.id === group.topicId)
 
@@ -41,8 +40,7 @@ const SingleGroupView = ({
                 group,
                 deleteFromGroupAction,
                 setSuccess,
-                setError,
-                clearNotifications
+                setError
               })
             }
             className={`edit-group-no__${groupId}__delete-instructor`}
@@ -83,8 +81,7 @@ const SingleGroupView = ({
                     student,
                     deleteFromGroupAction,
                     setSuccess,
-                    setError,
-                    clearNotifications
+                    setError
                   })
                 }
                 className={`edit-group-no__${
diff --git a/src/components/RegistrationDetailsPage.js b/src/components/RegistrationDetailsPage.js
index b01933b..748fe9a 100644
--- a/src/components/RegistrationDetailsPage.js
+++ b/src/components/RegistrationDetailsPage.js
@@ -163,6 +163,8 @@ class RegistrationDetailsPage extends React.Component {
       createdAt
     } = this.props.ownRegistration
 
+    const { peerReviewOpen } = this.props
+
     return (
       <div className="registration-details-container">
         <Typography variant="h4" gutterBottom>
@@ -172,7 +174,7 @@ class RegistrationDetailsPage extends React.Component {
           Registration date: {formatDate(createdAt)}
         </Typography>
 
-        <PeerReviewInfo />
+        {peerReviewOpen && <PeerReviewInfo />}
         <UserDetails student={student} />
         <PreferredTopics topics={preferred_topics} />
         <RegistrationAnswers questions={questions} />
@@ -183,7 +185,8 @@ class RegistrationDetailsPage extends React.Component {
 
 const mapStateToProps = (state) => {
   return {
-    ownRegistration: state.registration
+    ownRegistration: state.registration,
+    peerReviewOpen: state.registrationManagement.peerReviewOpen
   }
 }
 
diff --git a/src/reducers/actions/groupManagementActions.js b/src/reducers/actions/groupManagementActions.js
index 9009d15..407f332 100644
--- a/src/reducers/actions/groupManagementActions.js
+++ b/src/reducers/actions/groupManagementActions.js
@@ -13,9 +13,9 @@ const updateGroupTopicID = (topicID) => ({
   payload: topicID
 })
 
-const updateGroupInstructorID = (groupInstructorID) => ({
-  type: 'UPDATE_INSTRUCTOR_ID',
-  payload: groupInstructorID
+const updateGroupInstructor = (groupInstructor) => ({
+  type: 'UPDATE_INSTRUCTOR',
+  payload: groupInstructor
 })
 
 const updateGroupConfigurationID = (groupConfigurationID) => ({
@@ -43,9 +43,9 @@ const updateExistingGroup = (group) => ({
   payload: group
 })
 
-const deleteGroup = (group) => ({
+const deleteGroup = (groupId) => ({
   type: 'DELETE_GROUP',
-  payload: group
+  payload: groupId
 })
 
 const setUsers = (users) => ({
@@ -57,7 +57,7 @@ export default {
   updateCreateGroupFormName,
   updateStudentsForm,
   updateGroupTopicID,
-  updateGroupInstructorID,
+  updateGroupInstructor,
   updateGroupConfigurationID,
   createGroupSuccsess,
   setGroups,
diff --git a/src/reducers/groupPageReducer.js b/src/reducers/groupPageReducer.js
index 9f97630..1e6a533 100644
--- a/src/reducers/groupPageReducer.js
+++ b/src/reducers/groupPageReducer.js
@@ -2,7 +2,7 @@ const initialState = {
   groupName: '',
   students: '',
   groupTopicID: '',
-  groupInstructorID: '',
+  groupInstructor: null, // autosuggested user
   groupConfigurationID: '',
   groups: [],
   users: []
@@ -25,10 +25,10 @@ const groupPageReducer = (state = initialState, action) => {
       ...state,
       groupTopicID: action.payload
     }
-  case 'UPDATE_INSTRUCTOR_ID':
+  case 'UPDATE_INSTRUCTOR':
     return {
       ...state,
-      groupInstructorID: action.payload
+      groupInstructor: action.payload
     }
   case 'UPDATE_GROUP_CONFIGURATION_ID':
     return {
@@ -66,9 +66,7 @@ const groupPageReducer = (state = initialState, action) => {
   case 'DELETE_GROUP':
     return {
       ...state,
-      groups: state.groups.map((group) =>
-        group.id === action.payload.id ? '' : group
-      )
+      groups: state.groups.filter((group) => group.id !== action.payload)
     }
   case 'SET_USERS':
     return {
diff --git a/src/services/autocomplete.js b/src/services/autocomplete.js
new file mode 100644
index 0000000..89e8523
--- /dev/null
+++ b/src/services/autocomplete.js
@@ -0,0 +1,24 @@
+import axios from 'axios'
+import { BACKEND_API_BASE } from '../utils/config'
+import { getUserToken } from '../utils/functions'
+
+const url = `${BACKEND_API_BASE}/autocomplete`
+
+/**
+ * @returns {AutocompleteResult[]}
+ * @typedef {{ student_number: string, first_names: string, last_name: string }} AutocompleteResult
+ */
+const findUsersByPartialName = async (partialName) => {
+  // returns 400 if less than 3 characters passed!
+  const response = await axios.get(`${url}/users`, {
+    params: {
+      name: partialName
+    },
+    headers: { Authorization: 'Bearer ' + getUserToken() }
+  })
+
+  // [ { student_number, first_names, last_name } ]
+  return response.data
+}
+
+export default { findUsersByPartialName }
```
